server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,382 @@
+            }
+        } else {
+            listener.onFailure(new AlreadyClosedException("response already closed"));
+        }
+    }
+
+    private void restResponseFinished() {
+        assert Transports.assertTransportThread();
+        if (isRestResponseFinished.compareAndSet(false, true)) {
+            queueRefs.decRef();
+        }
+    }
+
+    private void drainQueue() {
+        assert isRestResponseFinished.get();
+        assert queueRefs.hasReferences() == false;
+        final var taskCount = queueLength.get() + 1;
+        final var releasables = new ArrayList<Releasable>(taskCount);
+        try {
+            if (releasable != null) {
+                releasables.add(releasable);
+                releasable = null;
+            }
+            ChunkedZipEntry entry;
+            while ((entry = entryQueue.poll()) != null) {
+                releasables.add(entry.releasable());
+            }
+            assert entryQueue.isEmpty() : entryQueue.size(); // no concurrent adds
+            assert releasables.size() <= taskCount : taskCount + " vs " + releasables.size();
+        } finally {
+            Releasables.closeExpectNoException(Releasables.wrap(releasables));
+        }
+    }
+
+    /**
+     * A {@link ChunkedRestResponseBodyPart} which will yield all currently-available chunks by consuming entries from the queue.
+     */
+    private final class QueueConsumer implements ChunkedRestResponseBodyPart {
+
+        private ZipEntry zipEntry;
+        private ChunkedRestResponseBodyPart bodyPart;
+        private boolean isPartComplete;
+        private boolean isLastPart;
+        private Consumer<ActionListener<ChunkedRestResponseBodyPart>> getNextPart;
+        private ArrayList<Releasable> nextReleasables = new ArrayList<>(); // preserved between chunks because will often be unused
+
+        QueueConsumer(ZipEntry zipEntry, ChunkedRestResponseBodyPart bodyPart) {
+            this.zipEntry = zipEntry;
+            this.bodyPart = bodyPart;
+        }
+
+        @Override
+        public boolean isPartComplete() {
+            return isPartComplete;
+        }
+
+        @Override
+        public boolean isLastPart() {
+            return isLastPart;
+        }
+
+        @Override
+        public void getNextPart(ActionListener<ChunkedRestResponseBodyPart> listener) {
+            assert getNextPart != null;
+            getNextPart.accept(listener);
+        }
+
+        private void transferReleasable(ArrayList<Releasable> releasables) {
+            assert queueRefs.hasReferences();
+
+            if (releasable == null) {
+                return;
+            }
+
+            if (releasables == nextReleasables) {
+                nextReleasables = new ArrayList<>();
+            }
+
+            releasables.add(releasable);
+            releasable = null;
+        }
+
+        @Override
+        public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
```

Comment:
This method is quite complex. I wonder whether there is a way to split or simplify it? IIUC, it deals with state changes for new entry (split between entry header and the rest of the entry), entry part continuation, entry continuation (immediate or delayed) and end of entries. Might be helpful to have separate methods to group them if possible?

Reply:
Fair point, the cold path is probably large enough to prevent JIT optimizations from kicking in.